File: MobileApp/Services/MqttService.cs
using System.Text.Json;
using HiveMQtt.Client;
using HiveMQtt.Client.Options;
using HiveMQtt.MQTT5.Types;
using Microsoft.Extensions.Options;
using ProiectVentilatie.Mobile.Models;

namespace ProiectVentilatie.Mobile.Services;

/// <summary>
/// Serviciu MQTT conectat la HiveMQ Cloud.
/// - Config via DI (IOptions&lt;MqttSettings&gt;)
/// - Reconnect cu exponential backoff (configurabil)
/// - Lifecycle: ConnectAsync/DisconnectAsync (OnSleep/OnResume)
/// - Subscribe: state (retained), online, event, log
/// - LastState cached — disponibil instant la connect
/// </summary>
public class MqttService : IMqttService, IDisposable
{
    private readonly MqttSettings _settings;
    private HiveMQClient? _client;
    private bool _disposed;
    private bool _intentionalDisconnect;
    private CancellationTokenSource? _reconnectCts;

    // Backoff reconnect — citit din settings
    private int _reconnectDelayMs;

    public event Action<VentilationState>? OnStateReceived;
    public event Action<bool>? OnConnectionChanged;
    public event Action<string>? OnOnlineStatusChanged;
    public event Action<string>? OnEventReceived;
    public event Action<string>? OnLogReceived;

    public VentilationState? LastState { get; private set; }
    public DateTime? LastStateReceivedAt { get; private set; }
    public bool IsConnected => _client?.IsConnected() ?? false;

    public MqttService(IOptions<MqttSettings> options)
    {
        _settings = options.Value;
        _reconnectDelayMs = _settings.ReconnectInitialMs;
    }

    public async Task ConnectAsync()
    {
        _intentionalDisconnect = false;
        _reconnectDelayMs = _settings.ReconnectInitialMs;

        if (_client != null && _client.IsConnected())
            return;

        _reconnectCts?.Cancel();
        _reconnectCts = new CancellationTokenSource();

        await ConnectInternalAsync();
    }

    public async Task DisconnectAsync()
    {
        _intentionalDisconnect = true;
        _reconnectCts?.Cancel();

        if (_client != null && _client.IsConnected())
        {
            try
            {
                await _client.DisconnectAsync();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[MQTT] Disconnect error: {ex.Message}");
            }
        }

        MainThread.BeginInvokeOnMainThread(() => OnConnectionChanged?.Invoke(false));
    }

    public async Task SendCommandAsync(object command)
    {
        if (_client == null || !_client.IsConnected())
        {
            Console.WriteLine("[MQTT] SendCommand skipped — not connected.");
            return;
        }

        try
        {
            var json = JsonSerializer.Serialize(command);
            await _client.PublishAsync(_settings.CommandTopic, json,
                QualityOfService.AtLeastOnceDelivery);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[MQTT] Send error: {ex.Message}");
        }
    }

    private async Task ConnectInternalAsync()
    {
        try
        {
            var options = new HiveMQClientOptions
            {
                Host = _settings.Host,
                Port = _settings.Port,
                UseTLS = true,
                UserName = _settings.Username,
                Password = _settings.Password,
                ClientId = $"maui-{Guid.NewGuid():N}".Substring(0, 23),
                CleanStart = true,
                KeepAlive = 60,
            };

            _client = new HiveMQClient(options);

            // Message handler — routing pe topic
            _client.OnMessageReceived += OnMessageReceived;

            // Disconnect handler — reconnect automat
            _client.AfterDisconnect += async (sender, args) =>
            {
                MainThread.BeginInvokeOnMainThread(() => OnConnectionChanged?.Invoke(false));

                if (!_intentionalDisconnect)
                {
                    Console.WriteLine($"[MQTT] Disconnected. Reconnect in {_reconnectDelayMs}ms...");
                    await ReconnectWithBackoffAsync();
                }
            };

            await _client.ConnectAsync();

            // Subscribe la toate topic-urile relevante.
            // State + Online sunt retained → primim imediat ultimele valori.
            await _client.SubscribeAsync(_settings.StateTopic,   QualityOfService.AtMostOnceDelivery);
            await _client.SubscribeAsync(_settings.OnlineTopic,  QualityOfService.AtLeastOnceDelivery);
            await _client.SubscribeAsync(_settings.EventTopic,   QualityOfService.AtMostOnceDelivery);
            await _client.SubscribeAsync(_settings.LogTopic,     QualityOfService.AtLeastOnceDelivery);

            _reconnectDelayMs = _settings.ReconnectInitialMs; // Reset backoff
            MainThread.BeginInvokeOnMainThread(() => OnConnectionChanged?.Invoke(true));
            Console.WriteLine("[MQTT] Connected + subscribed.");
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[MQTT] Connect error: {ex.Message}");
            MainThread.BeginInvokeOnMainThread(() => OnConnectionChanged?.Invoke(false));

            if (!_intentionalDisconnect)
            {
                await ReconnectWithBackoffAsync();
            }
        }
    }

    private async Task ReconnectWithBackoffAsync()
    {
        try
        {
            var token = _reconnectCts?.Token ?? CancellationToken.None;
            await Task.Delay(_reconnectDelayMs, token);

            // Exponential backoff
            _reconnectDelayMs = Math.Min(_reconnectDelayMs * 2, _settings.ReconnectMaxMs);

            if (!token.IsCancellationRequested)
            {
                await ConnectInternalAsync();
            }
        }
        catch (TaskCanceledException)
        {
            // Reconnect anulat — lifecycle DisconnectAsync
        }
    }

    private void OnMessageReceived(object? sender, HiveMQtt.Client.Events.OnMessageReceivedEventArgs args)
    {
        var topic = args.PublishMessage.Topic;
        var payload = args.PublishMessage.PayloadAsString;

        if (string.IsNullOrEmpty(payload)) return;

        try
        {
            if (topic == _settings.StateTopic)
            {
                var state = JsonSerializer.Deserialize<VentilationState>(payload);
                if (state != null)
                {
                    LastState = state;
                    LastStateReceivedAt = DateTime.Now;
                    MainThread.BeginInvokeOnMainThread(() => OnStateReceived?.Invoke(state));
                }
            }
            else if (topic == _settings.OnlineTopic)
            {
                MainThread.BeginInvokeOnMainThread(() => OnOnlineStatusChanged?.Invoke(payload));
            }
            else if (topic == _settings.EventTopic)
            {
                MainThread.BeginInvokeOnMainThread(() => OnEventReceived?.Invoke(payload));
            }
            else if (topic == _settings.LogTopic)
            {
                MainThread.BeginInvokeOnMainThread(() => OnLogReceived?.Invoke(payload));
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[MQTT] Parse error on {topic}: {ex.Message}");
        }
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _reconnectCts?.Cancel();
        _reconnectCts?.Dispose();

        if (_client != null)
        {
            _client.OnMessageReceived -= OnMessageReceived;
            try { _client.DisconnectAsync().Wait(2000); } catch { }
            _client.Dispose();
        }
    }
}

File: vitriinaAPP/js/mqtt-service.js
/**
 * MqttService — singleton MQTT WebSocket client for vitriina web app.
 * Connects to HiveMQ Cloud via WSS (port 8884).
 */
class MqttService {
  constructor() {
    this.client = null;
    this.isConnected = false;
    this.lastState = null;
    this.lastTvState = null;
    this.lastStateReceivedAt = null;
    this._listeners = {};

    this.settings = {
      host: '264f95b78b1d4733a57c7d0c6e045828.s1.eu.hivemq.cloud',
      port: 8884,
      username: 'ventilatie_app',
      password: 'AppVentil2024!',
      stateTopic:   'ventilatie/state',
      commandTopic: 'ventilatie/cmd',
      onlineTopic:  'ventilatie/online',
      eventTopic:   'ventilatie/event',
      logTopic:     'ventilatie/log',
      tvTopic:      'ventilatie/tv/state',
    };
  }

  // ── Event system ─────────────────────────────────
  on(event, handler) {
    if (!this._listeners[event]) this._listeners[event] = [];
    this._listeners[event].push(handler);
    return () => this.off(event, handler);
  }

  off(event, handler) {
    if (this._listeners[event])
      this._listeners[event] = this._listeners[event].filter(h => h !== handler);
  }

  _emit(event, data) {
    (this._listeners[event] || []).forEach(h => { try { h(data); } catch(e) {} });
  }

  // ── Connection ───────────────────────────────────
  connect() {
    if (this.client) return;
    const { host, port, username, password } = this.settings;
    const clientId = `vitriina_web_${Date.now().toString(16)}`;

    try {
      this.client = mqtt.connect(`wss://${host}:${port}/mqtt`, {
        clientId,
        username,
        password,
        clean: true,
        reconnectPeriod: 4000,
        connectTimeout: 12000,
        keepalive: 30,
      });
    } catch (e) {
      console.error('[MQTT] connect() failed:', e);
      return;
    }

    this.client.on('connect', () => {
      console.log('[MQTT] Connected');
      this.isConnected = true;
      this._emit('onConnectionChanged', true);

      const { stateTopic, onlineTopic, eventTopic, logTopic, tvTopic } = this.settings;
      this.client.subscribe([stateTopic, onlineTopic, eventTopic, logTopic, tvTopic], { qos: 1 }, (err) => {
        if (err) console.error('[MQTT] Subscribe error:', err);
      });

      // Ask ESP32 for fresh state
      this.sendCommand({ cmd: 'refresh' });
    });

    this.client.on('close', () => {
      console.log('[MQTT] Disconnected');
      this.isConnected = false;
      this._emit('onConnectionChanged', false);
    });

    this.client.on('reconnect', () => {
      this._emit('onConnectionChanged', false);
    });

    this.client.on('error', (err) => {
      console.error('[MQTT] Error:', err);
      this.isConnected = false;
      this._emit('onConnectionChanged', false);
    });

    this.client.on('message', (topic, payload) => {
      const str = payload.toString();
      const { stateTopic, onlineTopic, eventTopic, logTopic, tvTopic } = this.settings;

      if (topic === tvTopic) {
        try {
          const tv = JSON.parse(str);
          this.lastTvState = tv;
          this._emit('onTvStateReceived', tv);
        } catch (e) { console.error('[MQTT] tv parse fail:', e); }

      } else if (topic === stateTopic) {
        try {
          const state = JSON.parse(str);
          this.lastState = state;
          this.lastStateReceivedAt = new Date();
          this._emit('onStateReceived', state);
        } catch (e) { console.error('[MQTT] state parse fail:', e); }

      } else if (topic === onlineTopic) {
        this._emit('onOnlineStatusChanged', str);

      } else if (topic === eventTopic) {
        this._emit('onEventReceived', str);

      } else if (topic === logTopic) {
        try {
          const log = JSON.parse(str);
          this._emit('onLogReceived', log);
        } catch (e) {}
      }
    });
  }

  // ── Commands ─────────────────────────────────────
  sendCommand(payload) {
    if (!this.client || !this.isConnected) {
      console.warn('[MQTT] Not connected — command dropped:', payload);
      return Promise.resolve();
    }
    return new Promise((resolve, reject) => {
      this.client.publish(
        this.settings.commandTopic,
        JSON.stringify(payload),
        { qos: 1 },
        (err) => (err ? reject(err) : resolve())
      );
    });
  }

  disconnect() {
    if (this.client) {
      this.client.end(true);
      this.client = null;
      this.isConnected = false;
    }
  }
}

// Singleton exposed to all modules
window.mqttService = new MqttService();
